File: cpp/cpp_100-999/213.cc
// C++
// leetcode 213
// https://leetcode-cn.com/problems/house-robber-ii/
// 套用 198题的解函数，改名为 rob_one

class Solution {
public:
    int rob_one(vector<int>& nums) {
        int n = nums.size();
        if( n == 0)
            return 0;
        int dp[n + 1];
        // 装入0与1
        dp[0] = 0;
        dp[1] = nums[0];
        for(int i = 2; i <= n; ++i) {
            dp[i] = max(dp[i - 2] + nums[i - 1], dp[i - 1]);
        }
        return dp[n];
    }

    int rob(vector<int>& nums) {
        int size = nums.size();
        if(size== 0)
            return 0;
        if(size ==1)
            return nums[0];
        int last = nums[size-1];
        nums.pop_back();
        int withoutLast = rob_one(nums);
        nums.push_back(last);
        nums.erase(nums.begin());
        int withoutBegin = rob_one(nums);
        return max(withoutBegin , withoutLast);
    }
};

File: Go/Go_100-999/213.go
// C++
// leetcode 213
// https://leetcode-cn.com/problems/house-robber-ii/

func rob(nums []int) int {
	n:= len(nums)
	if n == 1 {
		return nums[0]
	}
	head := rob_one(nums[0:n-1])
	tail := rob_one(nums[1:])
	return max(head, tail)
}

func rob_one(nums []int) (res int) { // 198 的题解
	pre := make([]int, 3)
	getMaxPreFrom12 := func() int {
		res := 0
		for i := 1; i < 3; i++ {
			res = max(res, pre[i])
		}
		return res
	}

	updatePre := func(x int) {
		pre[2] = pre[1]
		pre[1] = pre[0]
		pre[0] = x
	}

	for _, x := range nums {
		cur := x + getMaxPreFrom12()
		res = max(res, cur)
		updatePre(cur)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}